src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt

```diff
@@ -0,0 +1,173 @@
+   limitations under the License.
+
+*/
+
+@file:OptIn(ExperimentalSerializationApi::class, InternalSerializationApi::class)
+
+package com.charleskorn.kaml
+
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.InternalSerializationApi
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.builtins.ListSerializer
+import kotlinx.serialization.builtins.MapSerializer
+import kotlinx.serialization.builtins.serializer
+import kotlinx.serialization.descriptors.PolymorphicKind
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.SerialKind
+import kotlinx.serialization.descriptors.buildSerialDescriptor
+import kotlinx.serialization.descriptors.nullable
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.encoding.encodeStructure
+
+internal object YamlNodeSerializer : YamlContentPolymorphicSerializer<YamlNode>(YamlNode::class) {
+    override val descriptor: SerialDescriptor = super.descriptor.nullable
+
+    override fun serialize(encoder: Encoder, value: YamlNode) {
+        encoder.asYamlOutput()
+        when (value) {
+            is YamlList -> encoder.encodeSerializableValue(YamlListSerializer, value)
+            is YamlMap -> encoder.encodeSerializableValue(YamlMapSerializer, value)
+            is YamlNull -> encoder.encodeSerializableValue(YamlNullSerializer, value)
+            is YamlScalar -> encoder.encodeSerializableValue(YamlScalarSerializer, value)
+            is YamlTaggedNode -> encoder.encodeSerializableValue(YamlTaggedNodeSerializer, value)
+        }
+    }
+
+    override fun deserialize(decoder: Decoder): YamlNode {
+        val input = decoder.asYamlInput<YamlInput>()
+        return if (input is YamlPolymorphicInput) YamlTaggedNode(input.typeName, input.node) else input.node
+    }
+
+    override fun selectDeserializer(node: YamlNode): DeserializationStrategy<YamlNode> {
+        error("implemented custom serialize logic")
+    }
+}
+
+internal object YamlScalarSerializer : KSerializer<YamlScalar> {
+    override val descriptor: SerialDescriptor =
+        PrimitiveSerialDescriptor("com.charleskorn.kaml.YamlScalar", PrimitiveKind.STRING)
+
+    override fun serialize(encoder: Encoder, value: YamlScalar) {
+        encoder.asYamlOutput()
+        try {
+            return encoder.encodeBoolean(value.toBoolean())
+        } catch (_: SerializationException) {
+        }
+        try {
+            return encoder.encodeLong(value.toLong())
+        } catch (_: SerializationException) {
+        }
+        try {
+            return encoder.encodeDouble(value.toDouble())
+        } catch (_: SerializationException) {
+        }
```

Comment:
Rather than this chain of `try` / `catch`, couldn't we inspect the value and determine the best option to use?

Reply:
I didn't want to copy the parsing code, to avoid duplicated code. Do you have an idea how to implement this without duplicating most of the `YamlScalar` code?
I mean `Boolean` has extra parsing code, integer values extra checks for octal, decimal and hexadecimal numbers, floating point numbers have extra checks for infinity and NaN. This was too much code for me to copy. Else the library gets inconsistent with future changes.

Reply:
What if we just always encode the value as a string?

Reply:
Then we would encode a double `1.2` as `"1.2"`, right? So the output might be unexpected to the user.

I could create `toBooleanOrNull`, `toByteOrNull`, …  functions in the `YamlScalar` class and use it in `toBoolean`, `toByte`, … functions and in the `YamlScalarSerializer`. This way the code doesn't get duplicated and we don't have to try catch everything.

Reply:
> I could create `toBooleanOrNull`, `toByteOrNull`, … functions in the `YamlScalar` class and use it in `toBoolean`, `toByte`, … functions and in the `YamlScalarSerializer`. This way the code doesn't get duplicated and we don't have to try catch everything.

This sounds good to me.

Reply:
done